src/main/java/org/genomicsdb/spark/GenomicsDBSchemaFactory.java

```diff
@@ -52,13 +53,101 @@
  * the vid map is not needed, such as from the scala api - then this is treated as a 
  * utility class.
  **/
-public class GenomicsDBSchemaFactory {
+public class GenomicsDBSchemaFactory implements JsonFileExtensions {
 
   private final Map<String, GenomicsDBVidSchema> vidMap;
 
   public GenomicsDBSchemaFactory(String loader){
     this.vidMap = buildVidSchemaMap(loader);
   }
+
+  public GenomicsDBSchemaFactory(VidMappingPB vidmapPB) {
+    this.vidMap = generateMapFromVidMap(vidmapPB);
+  }
+
+  private static Map<String, GenomicsDBVidSchema> generateMapFromVidMap(VidMappingPB vidmapPB) {
+    Map<String, GenomicsDBVidSchema> vidMap = new HashMap<String, GenomicsDBVidSchema>();
+    for (GenomicsDBVidMapProto.GenomicsDBFieldInfo f : vidmapPB.getFieldsList()) {
+      String key = f.getName();
+      Class<?> clazz = getFieldType(f.getType(0));
+      String length = getFieldLength(f);
+      if (f.getVcfFieldClassCount() == 1) {
+        vidMap.put((String) key, new GenomicsDBVidSchema(f.getVcfFieldClass(0).equals("INFO"), clazz, length));
+      }
+      // if field is both, add entries for INFO under it's name
+      // and then add an entry for FORMAT by adding the suffix _FORMAT
+      // mainly this is for DP
+      else {
+        vidMap.put((String) key, new GenomicsDBVidSchema(true, clazz, length));
```

Comment:
This was just refactoring existing code...but you're right that we don't really handle duplicate stuff well with this Spark dataframe api. We'll need to fix this...but also need to think about a bit more because we currently don't have a way to disambiguate field names if they're actually duplicates with the spark api....